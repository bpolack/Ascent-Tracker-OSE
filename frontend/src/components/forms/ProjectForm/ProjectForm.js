import React, { Component } from 'react'
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addProject, updateProject, getProject } from '../../../actions/project';

export class ProjectForm extends Component {
	constructor(props) {
		super(props);

		// Initial state
		this.state = {
			projectId: this.props.projectId,
			name: '',
			desc: ''
		}

		this.handleChange = this.handleChange.bind(this);
		this.handleSubmit = this.handleSubmit.bind(this);
	}

	async componentDidMount() {
		// Load the project details if projectId is present 
		const { projectId } = this.props;
		if (projectId) {
			await this.props.getProject(projectId);
			this.setState({
				name: this.props.project.project.name,
				desc: this.props.project.project.desc
			});
		}
	}

	handleChange(event) {
		event.preventDefault();
		this.setState({
			[event.target.name]: event.target.value,
		});
	}

	handleSubmit(event) {
		const { projectId, name, desc } = this.state;

		event.preventDefault();
		if (projectId) {
			this.props.updateProject({ id: projectId, name, desc });
		}
		else {
			this.props.addProject({ name, desc });
		}
		this.props.closeModal();
	}

	render() {

		let submitText = "Add Project";
		if (this.state.projectId) {
			submitText = "Save";
		}

		return (
			<form onSubmit={this.handleSubmit}>
				<label className="block mb-3">
					<span className="text-ascent-std">Project Name</span>
					<input type="text" name="name" onChange={this.handleChange} className="form-ascent" value={this.state.name} required />
				</label>
				<label className="block mb-3">
					<span className="text-ascent-std">Description</span>
					<textarea name="desc" onChange={this.handleChange} className="form-ascent h-32" value={this.state.desc} />
				</label>
				<input className="cursor-pointer leading-none mx-2 px-5 py-3 font-semibold transition-colors rounded-full text-white bg-indigo-500 hover:bg-pink-500" type="submit" value={submitText} />
			</form>
		)
	}
}

ProjectForm.propTypes = {
	getProject: PropTypes.func.isRequired,
	addProject: PropTypes.func.isRequired,
	updateProject: PropTypes.func.isRequired,
	project: PropTypes.object.isRequired
};

const mapStateToProps = (state) => ({
	project: state.project
});

export default connect(mapStateToProps, { getProject, addProject, updateProject })(ProjectForm);