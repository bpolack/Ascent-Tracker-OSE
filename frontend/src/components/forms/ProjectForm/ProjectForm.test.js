import React from 'react';
import { shallow } from 'enzyme';
import { ProjectForm } from './ProjectForm';

describe('ProjectForm', () => {
	it("Renders correctly without crashing", () => {
		const getProjectMock = jest.fn();
		const addProjectMock = jest.fn();
		const updateProjectMock = jest.fn();
		const project = {};

		const wrapper = shallow(
			<ProjectForm
				getProject={getProjectMock}
				addProject={addProjectMock}
				updateProject={updateProjectMock}
				project={project}
			/>
		);
		expect(wrapper).toMatchSnapshot();
	});

	describe('When user changes input', () => {
		it('form state is updated', () => {
			const getProjectMock = jest.fn();
			const addProjectMock = jest.fn();
			const updateProjectMock = jest.fn();
			const project = {};
			const preventDefault = jest.fn();

			const wrapper = shallow(
				<ProjectForm
					getProject={getProjectMock}
					addProject={addProjectMock}
					updateProject={updateProjectMock}
					project={project}
				/>
			);

			// Change form input values
			const nameElement = wrapper.find('input[name="name"]');
			const descElement = wrapper.find('textarea[name="desc"]');
			nameElement.simulate('change', { preventDefault, target: { name: 'name', value: 'Test Project' } });
			descElement.simulate('change', { preventDefault, target: { name: 'desc', value: 'Lorem Ipsum 123456' } });

			// Check that state has updated to match
			expect(wrapper.state('name')).toEqual("Test Project");
			expect(wrapper.state('desc')).toEqual("Lorem Ipsum 123456");
		});
	});

	describe('When user submits the form for a new project', () => {
		it('details are passed to addProject function', () => {
			const getProjectMock = jest.fn();
			const addProjectMock = jest.fn();
			const updateProjectMock = jest.fn();
			const project = {};
			const preventDefault = jest.fn();
			const closeModalMock = jest.fn();

			const wrapper = shallow(
				<ProjectForm
					getProject={getProjectMock}
					addProject={addProjectMock}
					updateProject={updateProjectMock}
					project={project}
					closeModal={closeModalMock} 
				/>
			);

			// Change form input values
			const formElement = wrapper.find('form');
			const nameElement = wrapper.find('input[name="name"]');
			const descElement = wrapper.find('textarea[name="desc"]');
			nameElement.simulate('change', { preventDefault, target: { name: 'name', value: 'Test Project' } });
			descElement.simulate('change', { preventDefault, target: { name: 'desc', value: 'Lorem Ipsum 123456' } });
			formElement.simulate('submit', { preventDefault });

			// Check that mock function called & with correct params
			expect(addProjectMock).toHaveBeenCalledTimes(1);
			expect(addProjectMock).toHaveBeenCalledWith({"name": "Test Project", "desc": "Lorem Ipsum 123456"});
		});
	});
});