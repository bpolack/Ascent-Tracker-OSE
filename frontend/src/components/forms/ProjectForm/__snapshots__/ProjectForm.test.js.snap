// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectForm Renders correctly without crashing 1`] = `ShallowWrapper {}`;
